### ISSUE TITLE
Regression in 2024-11-28: llvm-cov export errors with "truncated coverage data"

### ISSUE THREAD

**Issue by alex** [rust-lang/rust#133606]
### Code

Beginning with today's nightly, processing coverage produced by rustc binaries fails. Here is an example CI run: https://github.com/pyca/cryptography/actions/runs/12077661037/job/33681030705?pr=12066

The relevant snippet:

```
nox > /home/runner/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/llvm-cov export /home/runner/work/cryptography/cryptography/.nox/tests/lib/python3.12/site-packages/cryptography/hazmat/bindings/_rust.abi3.so -instr-profile=rust-cov.profdata '--ignore-filename-regex=[/\].cargo[/\]' '--ignore-filename-regex=[/\]rustc[/\]' '--ignore-filename-regex=[/\].rustup[/\]toolchains[/\]' '--ignore-filename-regex=[/\]target[/\]' --format=lcov
nox > Command /home/runner/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/llvm-cov export /home/runner/work/cryptography/cryptography/.nox/tests/lib/python3.12/site-packages/cryptography/hazmat/bindings/_rust.abi3.so -instr-profile=rust-cov.profdata '--ignore-filename-regex=[/\].cargo[/\]' '--ignore-filename-regex=[/\]rustc[/\]' '--ignore-filename-regex=[/\].rustup[/\]toolchains[/\]' '--ignore-filename-regex=[/\]target[/\]' --format=lcov failed with exit code 1:
error: failed to load coverage: '/home/runner/work/cryptography/cryptography/.nox/tests/lib/python3.12/site-packages/cryptography/hazmat/bindings/_rust.abi3.so': truncated coverage data
error: could not load coverage information
```

https://github.com/rust-lang/rust/pull/133418 is the only change in the regression range with a commit message mentioning coverage. cc: @Zalathar

Happy Thanksgiving!

### Version it worked on

`rustc 1.85.0-nightly (6b6a867ae 2024-11-27)`

### Version with regression

`rustc 1.85.0-nightly (a2545fd6f 2024-11-28)`

**Comment by Zalathar** [comment]
Hmm, my guess is that https://github.com/rust-lang/rust/pull/133418 is making it possible for *all* of a CGU's functions to be discarded at a late stage, so parts of the metadata end up being empty in a way that the `llvm-cov` isn't happy with.

Finding and testing a fix is probably going to be non-trivial (especially without a small repro), so for now it's probably wisest to revert.

**Comment by alex** [comment]
Regrettably, I don't have a reproducer smaller than "our entire large project with a bunch of deps" :-(

**Comment by Zalathar** [comment]
Yeah, minimizing this sort of coverage regression is annoying, so I generally don't expect issue reporters to do it themselves.

**Comment by arnodb** [comment]
Hi!

It's not minimized but https://github.com/arnodb/quirky_binder/actions/runs/12090701392/job/33718146447 faces exactly this problem.

It occurs with a build script. Unfortunately it's one of the last modules of the workspace to compile.

Commit `7c0ae33c25363b8b4637bb6501e10f2d2117f24c` crashes but it is also possible that other commits crash in the same way.

Hope this helps.

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/channel/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23133606.20Regression.20in.202024-11-28.3A.20llvm-cov.20export.20errors.20wi.E2.80.A6)).

@rustbot label -I-prioritize +P-medium

**Comment by Zalathar** [comment]
The revert has landed and seems to have fixed the regression, so closing as fixed.

(I'll need to be mindful of this failure when working on a new version of #133418.)

**Comment by Zalathar** [comment]
A new version of the original PR has been merged in #134497, and will be included in the next nightly.

Based on my testing, I believe that this issue should not reoccur, but let me know if something does go wrong.
